src/nnet3/nnet-simple-component.cc

```diff
@@ -668,12 +718,99 @@ void ClipGradientComponent::Backprop(const std::string &debug_info,
       in_deriv->ApplyCeiling(clipping_threshold_);
       in_deriv->ApplyFloor(-1 * clipping_threshold_);
     }
+
+    to_update->num_backproped_ += 1;
+    RepairGradients(debug_info, in_value, in_deriv, to_update);
   }
 }
 
+void ClipGradientComponent::RepairGradients(
+    const std::string &debug_info,
+    const CuMatrixBase<BaseFloat> &in_value,
+    CuMatrixBase<BaseFloat> *in_deriv, ClipGradientComponent *to_update) const {
+  KALDI_ASSERT(to_update != NULL);
+  BaseFloat default_clipped_proportion_threshold = 0.01;
+  BaseFloat default_clipping_to_value = 0.0;
+
+  // we use this 'repair_probability' (hardcoded for now) to limit
+  // this code to running on about half of the minibatches.
+  BaseFloat repair_probability = 0.5;
```

Comment:
In my BLSTM experiment, the self-repairs were called consistently at least between jobs within the same training iteration:
train.3.1 1391 times out of 12180
train.3.2 1200 times out of 12180
train.3.3 1372 times out of 12180
train.5.1 1639 times out of 12400
train.5.2 1224 times out of 12180
train.5.3 1344 times out of 12180
train.6.3   600 times out of 12240
train.8.2   432 times out of 12180

Perhaps there are two opposite forces when self-repair: one reducing the magnitude of gradient caused by reducing the magnitudes of parameter values, and the other one increasing the magnitude of gradient caused by reducing the inputs to nonlinearity on the path. It seems the former one dominates in this specific experiment.